sourcegen-generator/src/main/java/io/micronaut/sourcegen/generator/visitors/DelegateAnnotationVisitor.java

```diff
@@ -0,0 +1,251 @@
+                    );
+                } catch (Exception e) {
+                    throw new ProcessingException(element, "Failed to generate a delegate: " + e.getMessage(), e);
+                }
+            });
+        } catch (ProcessingException e) {
+            throw e;
+        } catch (Exception e) {
+            SourceGenerators.handleFatalException(
+                element,
+                Delegate.class,
+                e,
+                (exception -> {
+                    processed.remove(element.getName());
+                    throw exception;
+                })
+            );
+        }
+    }
+
+    private Collection<ClassElement> getAllInterfaces(ClassElement element) {
+        Queue<ClassElement> interfaces = new LinkedList<>();
+        Set<ClassElement> visited = new HashSet<>();
+        interfaces.addAll(element.getInterfaces());
+
+        while (!interfaces.isEmpty()) {
+            ClassElement current = interfaces.poll();
+            visited.add(current);
+            for (ClassElement iface : current.getInterfaces()) {
+                if (!visited.contains(iface)) {
+                    interfaces.add(iface);
+                }
+            }
+        }
+        return visited;
+    }
+
+    private void addDelegateMethods(
+            ClassElement element, ClassDefBuilder builder,
+            TypeDef typeDef, TypeDef delegateType, Set<MethodSignature> createdMethods
+    ) {
+        for (MethodElement method: element.getMethods()) {
+            if (method.isPrivate()) {
+                continue;
+            }
+            MethodSignature signature = new MethodSignature(
+                method.getName(),
+                Arrays.stream(method.getParameters()).map(v -> v.getType().getName()).toList()
+            );
+            if (createdMethods.contains(signature)) {
+                continue;
+            }
+            createdMethods.add(signature);
+
+            TypeDef returnType = TypeDef.of(method.getGenericReturnType());
+            MethodDefBuilder methodBuilder = MethodDef.builder(method.getName())
+                .overrides()
+                .returns(returnType);
+            List<ExpressionDef> parameters = new ArrayList<>();
+            if (method.isPublic()) {
+                methodBuilder.addModifiers(Modifier.PUBLIC);
+            } else if (method.isProtected()) {
+                methodBuilder.addModifiers(Modifier.PROTECTED);
+            }
+            for (ParameterElement parameter: method.getParameters()) {
+                ParameterDef def = ParameterDef.builder(
+                    parameter.getName(), TypeDef.of(parameter.getGenericType())
+                ).build();
+                methodBuilder.addParameter(def);
+                parameters.add(def.asVariable());
+            }
+            ExpressionDef.CallInstanceMethod delegateCall = ExpressionDef.invoke(new VariableDef.Field(
+                new This(delegateType), DELEGATEE_MEMBER, typeDef
```

Comment:
use build method to access `this`

Reply:
I am not sure what you mean. I can see no alternative method for creating `this`.

Reply:
build method has a lamda that provides an access to this and parameters

Reply:
The best I could find is
```java
ExpressionDef.CallInstanceMethod delegateCall = new This(delegateType)
                .field(DELEGATEE_MEMBER, typeDef)
                .invoke(method.getName(), returnType, parameters);
```

Reply:
Use `.build((self, parameterDefs) ->`